File: go/424.go
package main

func main() {
	println(characterReplacement("ABAB", 2))
	println(characterReplacement("AABABBA", 1))
}

func characterReplacement(s string, k int) int {
	counter := [26]int{}
	res := 0
	l := 0
	for r := range s {
		counter[s[r]-'A']++

		if (r-l+1)-maxCounter(counter) <= k {
			res = max(res, r-l+1)
		}
		for ; (r-l+1)-maxCounter(counter) > k; l++ {
			counter[s[l]-'A']--
		}
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxCounter(counter [26]int) int {
	res := counter[0]
	for _, v := range counter {
		res = max(res, v)
	}
	return res
}

File: py/424.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:        
        counter = {}
        res = 0
        l = 0
        for r in range(len(s)):
            c = s[r]
            counter[c] = 1 + counter.get(c, 0)
            if r - l + 1 - max(counter.values()) <= k:
                res = max(res, r - l + 1)
            while r - l + 1 - max(counter.values()) > k:
                c = s[l]
                counter[c] -= 1 
                l += 1
        
        return res
				
        
print(Solution().characterReplacement('ABAB', 2))
print(Solution().characterReplacement('AABABBA', 1))
        
